contracts/0.8.9/oracle/AccountingOracle.sol

```diff
@@ -700,174 +710,145 @@ contract AccountingOracle is BaseOracle {
-            iter.lastModuleId = int256(moduleId);
-            iter.lastNodeOpId = int256(nodeOpId);
-        }
+    function _processExtraDataItems(bytes calldata data, ExtraDataIterState memory iter) internal {
+        uint256 dataOffset = iter.dataOffset;
+
+        /// @solidity memory-safe-assembly
+        while (dataOffset < data.length) {
+            uint256 index;
+            uint256 itemType;
+
+            assembly {
+                // layout at the dataOffset:
+                // |  3 bytes  | 2 bytes  |   X bytes   |
+                // | itemIndex | itemType | itemPayload |
+                let header := calldataload(add(data.offset, dataOffset))
+                index := shr(232, header)
+                itemType := and(shr(216, header), 0xffff)
+                dataOffset := add(dataOffset, 5)
+            }
 
-        while (iter.nextIndex < items.length) {
-            iter.nopIds.clear();
-            iter.keyCounts.clear();
-
-            _processSingleModule(iter, items);
-
-            if (iter.lastType != -1) {
-                assert(iter.lastModuleId >= 0);
-                assert(iter.lastNodeOpId >= 0);
-
-                if (iter.lastType == int256(EXTRA_DATA_TYPE_STUCK_VALIDATORS)) {
-                    stakingRouter.reportStakingModuleStuckKeysCountByNodeOperator(
-                        uint256(iter.lastModuleId),
-                        iter.nopIds.pointer(),
-                        iter.keyCounts.pointer()
-                    );
-                } else if (iter.lastType == int256(EXTRA_DATA_TYPE_EXITED_VALIDATORS)) {
-                    stakingRouter.reportStakingModuleExitedKeysCountByNodeOperator(
-                        uint256(iter.lastModuleId),
-                        iter.nopIds.pointer(),
-                        iter.keyCounts.pointer()
-                    );
-                } else {
-                    revert UnsupportedExtraDataType(uint256(iter.lastType));
+            if (!iter.started) {
+                if (index != 0) {
+                    revert UnexpectedExtraDataIndex(0, index);
                 }
+                iter.started = true;
+            } else if (index != iter.index + 1) {
+                revert UnexpectedExtraDataIndex(iter.index + 1, index);
             }
-        }
 
-        return iter.nextIndex == 0 ? 0 : items[iter.nextIndex - 1];
-    }
+            iter.index = index;
+            iter.itemType = itemType;
+            iter.dataOffset = dataOffset;
 
-    function _processSingleModule(ExtraDataIterState memory iter, uint256[] calldata items)
-        internal pure
-    {
-        if (items.length == 0) {
-            return;
-        }
+            if (itemType == EXTRA_DATA_TYPE_EXITED_VALIDATORS ||
+                itemType == EXTRA_DATA_TYPE_STUCK_VALIDATORS
+            ) {
+                _processExtraDataItem(data, iter);
+            } else {
+                revert UnsupportedExtraDataType(index, itemType);
+            }
```

Comment:
`_processExtraDataItem` handles specific two item types, this may not be flexible enough. Not sure if solidity allows it, but something like this might be more flexible and improve the semantics:

```solidity
if (itemType == EXTRA_DATA_TYPE_EXITED_VALIDATORS) {
    (uint256 moduleId, bytes calldata nodeOpIds, bytes calldata valsCounts) = _extractKeysCounts(data, iter);
    IStakingRouter(iter.stakingRouter)
        .reportStakingModuleExitedKeysCountByNodeOperator(moduleId, nodeOpIds, valsCounts);
} else if (itemType == EXTRA_DATA_TYPE_STUCK_VALIDATORS) {
    (uint256 moduleId, bytes calldata nodeOpIds, bytes calldata valsCounts) = _extractKeysCounts(data, iter);
    IStakingRouter(iter.stakingRouter)
        .reportStakingModuleStuckKeysCountByNodeOperator(moduleId, nodeOpIds, valsCounts);
} else {
    revert UnsupportedExtraDataType(index, itemType);
}
```

1) rename `_processExtraDataItem` into something more specific that performs the key extraction, like `_extractKeysCounts`
2) move Staking Router methods calls from `_extractKeysCounts` here

Reply:
yeah, this structuring would be better, and Solidity allows this. I'll try finding time to improve as you suggested